### PR TITLE
luminous: os/bluestore: avoid overhead of std::function in blob_t.

### PR DESCRIPTION
**Pull request by rzarzynski**
This is luminous backport of PR #20294. The extra commit that brings the compliance with C++11's `static_assert()` is only to make reviewing easier. I plan to squash it afterward. Apart from that there is no other change.

### CODE REVIEW

File: src/os/bluestore/bluestore_types.h [hunk 1]
```diff
@@ -729,8 +730,11 @@ struct bluestore_blob_t {
     }
   }
 
-  int map(uint64_t x_off, uint64_t x_len,
-	   std::function<int(uint64_t,uint64_t)> f) const {
+  template<class F>
+  int map(uint64_t x_off, uint64_t x_len, F&& f) const {
+    static_assert(std::is_invocable_r_v<int, F, uint64_t, uint64_t>,
```
Comment (tchaikov) [inline src/os/bluestore/bluestore_types.h]: please note `std::is_invocable_r_v<>` was introduced in C++17. while luminous is using C++14.
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/os/bluestore/bluestore_types.h: In member function ‘int bluestore_blob_t::map(uint64_t, uint64_t, F&&) const’:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/os/bluestore/bluestore_types.h:735:19: error: ‘is_invocable_r_v’ is not a member of ‘std’
     static_assert(std::is_invocable_r_v<int, F, uint64_t, uint64_t>,
                   ^
```

Comment (rzarzynski) [reply src/os/bluestore/bluestore_types.h]: Yeah, the same with `std::is_invocable`. Just pushed a version with the asserts removed.

**Comment by yuriw** [comment]
[wip-yuri2-testing-2018-04-06-1946-luminous](https://trello.com/c/jnS4VFcV/287-wip-yuri2-testing-2018-04-06-1946-luminous)

**Comment by yuriw** [comment]
[wip-yuri4-testing-2018-04-09-1710-luminous](https://trello.com/c/yyDKpeL2/290-wip-yuri4-testing-2018-04-09-1710-luminous)

**Pull request merged**